src/emitters/spot.cpp

```diff
@@ -0,0 +1,176 @@
+
+        // 2. Sample directional component
+        const Transform4f &trafo = m_world_transform->eval(time, active);
+        Vector3f d = trafo * warp::square_to_uniform_cone(spatial_sample, m_cos_cutoff_angle);
+
+        return { Ray3f(trafo * Point3f(0.f), d, time, wavelengths),
+                select(m_beam_width == m_cutoff_angle,
+                        unpolarized<Spectrum>(spec_weight) * 2 * math::Pi<Float> * (1 - m_cos_cutoff_angle),
+                        unpolarized<Spectrum>(spec_weight) * (2 * math::Pi<Float> * (1 - (sin(m_cutoff_angle) - sin(m_beam_width))/(m_cutoff_angle - m_beam_width)))) };
+    }
+
+    std::pair<DirectionSample3f, Spectrum> sample_direction(const Interaction3f &it,
+                                                            const Point2f & /*sample*/,
+                                                            Mask active) const override {
+        MTS_MASKED_FUNCTION(ProfilerPhase::EndpointSampleDirection, active);
+
+        auto trafo = m_world_transform->eval(it.time, active);
+
+        DirectionSample3f ds;
+        ds.p     = trafo.translation();
+        ds.n     = 0.f;
+        ds.uv    = 0.f;
+        ds.time  = it.time;
+        ds.pdf   = 1.f;
+        ds.delta = true;
+        ds.object = this;
+        ds.d     = (ds.p - it.p);
+        ds.dist  = norm(ds.d);
+        Float inv_dist = rcp(ds.dist);
+        ds.d *= inv_dist;
+
+        Spectrum spec = falloff_curve(ds.d, it.time, it.wavelengths, active) * (inv_dist * inv_dist);
+        return { ds, unpolarized<Spectrum>(spec) };
+    }
+
+    Float pdf_direction(const Interaction3f &, const DirectionSample3f &, Mask) const override {
+        return 0.f;
+    }
+
+    Spectrum eval(const SurfaceInteraction3f &, Mask) const override { return 0.f; }
+
+    ScalarBoundingBox3f bbox() const override {
+        return m_world_transform->translation_bounds();
+    }
+
+    void traverse(TraversalCallback *callback) override {
+        callback->put_object("intensity", m_intensity.get());
+    }
+
+    std::string to_string() const override {
+        std::ostringstream oss;
+        oss << "SpotLight[" << std::endl
+            << "  world_transform = " << string::indent(m_world_transform->to_string()) << "," << std::endl
+            << "  intensity = " << m_intensity << "," << std::endl
+            << "  cutoff_angle = " << m_cutoff_angle << "," << std::endl
+            << "  beam_width = " << m_beam_width << "," << std::endl
+            << "  texture = " << (m_texture ? string::indent(m_texture->to_string()) : "")
+            << "  medium = " << (m_medium ? string::indent(m_medium->to_string()) : "")
+            << "]";
+        return oss.str();
+    }
+
+    MTS_DECLARE_CLASS()
+private:
+    ref<Texture> m_intensity;
+	ref<Texture> m_texture;
+	Float m_beam_width, m_cutoff_angle, m_uv_factor;
+	Float m_cos_beam_width, m_cos_cutoff_angle, m_inv_transition_width;
```

Comment:
Those members should be `ScalarFloat` instead of `Float`.

For variants other than `scalar_*`, `Float` represent a dynamic array of `float` values. Here a spot light object only represent a single spot light so those members should be scalar.

Reply:
This being said, if the plan is to be able to optimize those parameters using autodiff, then they should indeed be `Float`.

Reply:
I think we should let these be optimizable. Is there any reason we should not?

Reply:
For now let's have them as `ScalarFloat` for simplicity. Other plugins also have tons of properties that are not optimizable ATM, we will do a pass over all of those in the future.

Reply:
Converting `m_cos_cutoff_angle` to `ScalarFloat` causes `square_to_uniform_cone` to complain. Should I leave just `m_cos_cutoff_angle` to `Float`?

Reply:
No, you can simply cast it as a ` Float` when calling the function.